### ISSUE TITLE
C++ header with long lines hangs VSCode

### ISSUE THREAD

**Issue by matter123** [jeff-hykin/cpp-textmate-grammar#415]
**From**: https://github.com/microsoft/vscode/issues/84723#issue-522387626
**By**: @maroc81
Issue Type: **Bug**

Loading a C++ header with long lines will cause VSCode to hang and show the "not responding" dialog.

To reproduce, create a header file with the contents such as this: https://gist.github.com/maroc81/f80bcecbd12e3b9a0496d18b970e880e

With just one line in the file, VSCode is slow but does not hang. If you duplicate the line (a hundred times or so), then trying to open the file will hang for some time. Eventually the file contents will show but VSCode is mostly unusable until you close the file.

I tested on a machine with an older version of VSCode (version 1.29.0) and it had no problems loading the file.

VS Code version: Code 1.40.0 (`86405ea`, 2019-11-06T18:14:08.920Z)
OS version: Linux x64 5.0.0-32-generic

Extensions disabled

**Comment by matter123** [comment]
<details>
<summary>Perf Results</summary>
> node test/source/index.js report "perf" "/Users/mfosdick/slow.cpp"

/Users/mfosdick/slow.cpp
source.cpp
source.cpp.embedded.macro
requested grammar outside of this repository
requested grammar outside of this repository
total time: 5565.513792991638ms

match failure
pattern name             average time    total time
source.cpp:456           149.223373     4625.925
source.cpp:11866          26.565495      504.744
source.cpp:3947            0.165387        3.142
source.cpp:6895            0.084275        2.613
source.cpp:10092           0.015942        0.797
source.cpp:9994            0.014005        0.084
source.cpp:10216           0.012356        0.618
source.cpp:10791           0.011677        0.362
source.cpp:3676            0.009671        0.300
source.cpp:10134           0.009581        0.297
source.cpp:304             0.009565        0.182
source.cpp:5928            0.009548        0.296
source.cpp:7896            0.009343        0.290
source.cpp:7908            0.009308        0.289
source.cpp:5247            0.009101        0.282
source.cpp:2108            0.009052        0.281
source.cpp:6280            0.009048        0.280
source.cpp:13634           0.009043        0.280
source.cpp:12053           0.009024        0.280
source.cpp:6235            0.008997        0.279

matched, not chosen
pattern name             average time    total time
source.cpp:9977            0.034787        0.209
source.cpp:3933            0.030237        0.181
source.cpp:11424           0.019284        0.116

matched, chosen
pattern name             average time    total time
source.cpp:3873            0.048125        0.289
source.cpp:9983            0.025479        0.153
source.cpp:3933            0.014848        0.089
source.cpp:11424           0.011349        0.068
source.cpp:9977            0.010068        0.060
</details>

line 456 is `#builtin_storage_type_initilizer`

**Comment by matter123** [comment]
@jeff-hykin It looks like std_space is no longer preventing backtracing
the generated regex is `((?:(?:(?>\\s+)|(\\/\\*)((?>(?:[^\\*]|(?>\\*+)[^\\/])*)((?>\\*+)\\/)))+...` note the lack of `\\s++`

**Comment by matter123** [comment]
Making the entire std_space pattern atomic fixes some issue (with a minor regression in `issues/027.cpp`).

New perf report
```
> node test/source/index.js report "perf" "/Users/mfosdick/slow.cpp"

/Users/mfosdick/slow.cpp
source.cpp
source.cpp.embedded.macro
requested grammar outside of this repository
requested grammar outside of this repository
total time: 1456.5780340135098ms

match failure
pattern name             average time    total time
source.cpp:11866          26.816601      509.515
source.cpp:456            16.789181      520.465
source.cpp:3947            0.119540        2.271
source.cpp:6895            0.083969        2.603
source.cpp:9994            0.015110        0.091
source.cpp:10092           0.014871        0.744
source.cpp:304             0.011362        0.216
source.cpp:10791           0.011068        0.343
source.cpp:12014           0.009505        0.295
source.cpp:3067            0.009380        0.469
source.cpp:6433            0.009346        0.290
source.cpp:6280            0.009345        0.290
source.cpp:10037           0.009219        0.461
source.cpp:7923            0.009215        0.286
source.cpp:13007           0.009192        0.285
source.cpp:5247            0.009130        0.283
source.cpp:5809            0.009127        0.283
source.cpp:10134           0.009065        0.281
source.cpp:7912            0.009060        0.281
source.cpp:10555           0.009041        0.280

matched, not chosen
pattern name             average time    total time
source.cpp:9977            0.036077        0.216
source.cpp:3933            0.033127        0.199
source.cpp:11424           0.020033        0.120

matched, chosen
pattern name             average time    total time
source.cpp:3873            0.050650        0.304
source.cpp:9983            0.027336        0.164
source.cpp:3933            0.016721        0.100
source.cpp:9977            0.011374        0.068
source.cpp:11424           0.009928        0.060
```

line 11866 is static assert.

**Comment by matter123** [comment]
Alternatively, at the risk of making `std_space.then(std_space)` not match, `std_space` could assert that there is a non-space (or boundary) immdedialty before it matches.

**Comment by jeff-hykin** [comment]
Thanks for jumping on this.

What causes `issues/027.cpp` to fail?

> Removing all the leading std_space and just having a common std_space pattern in every context to eat characters might help.

This sounds like a good idea but I think sometimes its difficult because patterns need a lookbehind as in `lookbehind().then(spaces).then(stuff)` but then that lookbehind causes it to take priority over other patterns

**Comment by matter123** [comment]
The issue was that std_space was lazy. This meant it matched /**/ but not the following space, and it was unable to backtrack to correct that. Making std_space greedy fixed the issue without causing test failure.

**Comment by jeff-hykin** [comment]
Can we wrap the atomic group inside of a optional group to keep the conditionality?

**Comment by matter123** [comment]
> >Removing all the leading std_space and just having a common std_space pattern in every context to eat characters might help.
>
>This sounds like a good idea but I think sometimes its difficult because patterns need a lookbehind as in lookbehind().then(spaces).then(stuff) but then that lookbehind causes it to take priority over other patterns

Yep, thats why I just went with adding a boundry condition to matching spaces.

**Comment by matter123** [comment]
> Can we wrap the atomic group inside of a optional group to keep the conditionality?

Sorry, I am not sure what you mean by the conditionality.

**Comment by jeff-hykin** [comment]
> > Can we wrap the atomic group inside of a optional group to keep the conditionality?
> 
> Sorry, I am not sure what you mean by the conditionality.

Have it still work with the `/**/` case (no spaces).

**Comment by matter123** [comment]
It does.

**Comment by jeff-hykin** [comment]
Ah okay that makes more sense

**Comment by matter123** [comment]
@maroc81 This has been fixed in version 1.14.14 of the "Better C++ Syntax" extension.
